tensor_comprehensions/tc_unit.py

```diff
@@ -0,0 +1,674 @@
+            logger.info('Backwards autotuning cache will be saved to: {}.cuda/options'.format(cache_file))
+        kwargs["type"] = "backward"
+        options = get_options_from_kwargs_and_tuner_cache(backward_name, cache_file, options_cache, *inputs, **kwargs)
+        backward_best_options = self.tune_and_store(
+            backward_name, inputs, mapping_options=options, cache_file=cache_file
+        )
+        # update the cache with the options
+        options_cache["backward"] = backward_best_options
+        self.tuner_cache[hash_key] = options_cache
+        return [forward_best_options, backward_best_options]
+
+
+###############################################################################
+# TC engine - ATen based
+###############################################################################
+class TcCompilationUnit(object):
+
+    def __init__(self):
+        self.cu = ATenCompilationUnit()
+        self.tc_lang = None
+        self.compilation_cache = {}
+
+    def define(self, tc_lang):
+        self.tc_lang = tc_lang
+        self.cu.define(tc_lang)
+
+    # we could have multiple TC definitions and want to run one of them
+    def compile(self, name, inputs, **kwargs):
+        # append the language so that we can use it for creating Autotuner object
+        # to load options cache
+        kwargs["tc_lang"] = self.tc_lang
+        if "type" not in kwargs:
+            kwargs["type"] = 'forward'
+        options = get_options_from_kwargs(name, *inputs, **kwargs)
+        handle = self.cu.compile(name, inputs, options)
+        return handle
+
+    def run(self, handle, name, inputs, **kwargs):
+        outputs = []
+        if "outputs" in kwargs and kwargs["outputs"] is not None:
+            outputs = kwargs["outputs"]
+            if not isinstance(outputs, list):
+                outputs = [outputs]
+        self.cu.run(name, inputs, outputs, handle)
+        return outputs
+
+    def compile_and_run(self, name, inputs, **kwargs):
+        handle = self.compile(name, inputs, **kwargs)
+        return self.run(handle, name, inputs, **kwargs)
+
+    def manual_cuda_injection(
+        self, name, injected_kernel_name, cuda_code, inputs, grid, block
+    ):
+        self.cu.inject_cuda(
+            name, injected_kernel_name, cuda_code, inputs, grid, block
+        )
+
+
+###############################################################################
+# User Facing Proxy object
+###############################################################################
+class TcUnit(object):
+
+    def __init__(self, lang, **kwargs_define):
+        self.cu = TcCompilationUnit()
+        self.cu.define(lang)
+        self.kwargs_define = kwargs_define
+        self.lang = lang
+        self.tuner = None   # this tuner maintains a cache for kernels/input sizes tuned so far
+
+    def __call__(self, *inputs, **kwargs):
+        r"""Runs the define TC language on given inputs.
+
+        Args:
```

Comment:
nit: there seems to be a lot of duplicate between the ```__call__``` and the ```autotune``` comments, is there a way to compress a bit?